platform/pwl.h

```diff
@@ -18,9 +41,192 @@ typedef struct {
     struct wl_compositor *compositor;
 } PwlDisplay;
 
+typedef struct {
+    struct egl_display *display;
+    EGLContext context;
+    EGLConfig egl_config;
+} PwlEGLData;
+
+typedef struct {
+    struct xkb_context* context;
+    struct xkb_keymap* keymap;
+    struct xkb_state* state;
+
+    struct xkb_compose_table* compose_table;
+    struct xkb_compose_state* compose_state;
+
+    struct {
+        xkb_mod_index_t control;
+        xkb_mod_index_t alt;
+        xkb_mod_index_t shift;
+    } indexes;
+
+    struct {
+        xkb_mod_index_t control;
+        xkb_mod_index_t alt;
+        xkb_mod_index_t shift;
+    } modifier;
+
+    uint8_t modifiers;
+} PwlXKBData;
+
+#if HAVE_DEVICE_SCALING
+typedef struct output_metrics {
+  struct wl_output *output;
+  int32_t name;
+  int32_t scale;
+} output_metrics;
```

Comment:
The device scaling should be tracked for each surface, because each Wayland surface (“window”) can be in a different output, and outputs can have each a different device scaling factor.